specs/StartInBackgroundMode.md

```diff
@@ -0,0 +1,48 @@
+StartInBackgroundMode
+===
+
+# Background
+By default, WebView2 instances get created with some priorities set higher, assuming the WV2 is being created in the foreground for an important purpose. However, not every developer wants to boost the WebView2 prelaunch phase at the cost of utilizing more system resources. The `StartInBackgroundMode` property specifies that whether WebView2 created with launch boost or not.
+
+
+# Description
+Setting `StartInBackgroundMode` to TRUE will signal WebView2 to start in a 'background mode' which will attempt to use less system resources during the creation and initial load. Default is FALSE.
+
+# Examples
+## Win32 C++
+```cpp
+auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();
+options->put_StartInBackgroundMode(TRUE));
+HRESULT hr = CreateCoreWebView2EnvironmentWithOptions(
+  nullptr, m_userDataFolder.c_str(), options.Get(),
+  Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>(
+     this, &AppWindow::OnCreateEnvironmentCompleted).Get());
+```
+## WinRT and .NET
+```c#
+auto options = new CoreWebView2EnvironmentOptions();
+options.StartInBackgroundMode = true;
+auto environment = await CoreWebView2Environment.CreateAsync(BrowserExecutableFolder, UserDataFolder, options);
+```
+
+
+# API Details
+## Win32 C++
+```cpp
+interface ICoreWebView2StagingEnvironmentOptions3 : IUnknown 
+{
+  /// Setting `StartInBackgroundMode` to `TRUE` will signal WebView2 to start in a 'background mode'
+  /// which will attempt to use less system resources during the creation and initial load.
+  /// Default is FALSE.
+  [propget] HRESULT StartInBackgroundMode([out, retval] BOOL* value);

```

Comment:
Saying 'start in background mode' makes it sound like WebView2 has entered this mode and its not clear how or when it will exit. So rather than 'start in .. mode' I'd rather something that ties it specifically to when this low priority behavior will apply. Like ShouldCreateWithLowPriority or ShouldCreateWithMinimalResourceUsage or something?

Reply:
Also the term 'background' is a bit ambiguous as well. It has some ties to UI when this feature is really about resource usage / priority.

Reply:
Yep. I agree we should use something like `ShouldCreateWithLowPriority` or `ShouldCreateWithMinimalResourceUsage` 

From the [reference](https://microsoft.sharepoint.com/:w:/r/teams/Edge/_layouts/15/Doc.aspx?sourcedoc=%7B8C776052-C93A-4EE1-8658-2D8182757B19%7D&file=Launch%20Priority%20Announcement%20-%20August%202021.docx&action=default&mobileredirect=true&share=IQFSYHeMOsnhToZYLYGCdXsZAcFBNjNDc_-s7rGB6eXHRaw&cid=7a8fc950-9a02-45a5-a8b4-7f1d6a78038c&wdLOR=c4DF38799-7315-4F39-87AF-F4ABB139DFAD)
> When an application using WebView2 is launching, there’s a period of time before foreground boost kicks-in and offers benefits. During this period WV2 processes are susceptible to starvation. We’re adding WV2 functionality that will ship in Edge 94 that __raises the launch priority of WV2 processes to ensure that WV2 processes are given higher CPU priority and access to resources.__ This will give the manager process higher CPU priority until the application has a foreground process that will trigger foreground boost. 

Setting this property to `TRUE` just disables `launch_boost` which actively raises the priority. The code inside [launch_win.cc](https://microsoft.visualstudio.com/Edge/_git/chromium.src?path=%2Fbase%2Fprocess%2Flaunch_win.cc&_a=contents&version=GBmain)(Line 442). 

So i'm not able to make choice around `ShouldCreateWithLowPriority` or something like  `ShouldCreateWithoutRaisingPriority`.

Reply:
I would bet foreground boost won't exist on other OSes. Maybe we can make this an enum property instead and name it CreationPriority = Default/High/Normal and say in the docs this is best effort

Reply:
> I would bet foreground boost won't exist on other OSes. 

Yep. Only Win. Good point! I will add description for this.

> Maybe we can make this an enum property instead and name it CreationPriority = Default/High/Normal and say in the docs this is best effort

Good alternative! 
__Default, High__ represents create with trying to Enable launch boost.
  
One thing bothers me is this property is not the only signal determines if we create WV2 with higher Priority. `CanLaunchManagerProcessWithBoost()` also contribute. 
```c++ 
bool use_launch_boost =
      CanLaunchManagerProcessWithBoost() && !(creationPriority == CreationPriority.Normal);
TRACE_EVENT_INSTANT1("edge_webview", "WebView2_ManagerProcessLaunchBoost",
                       TRACE_EVENT_SCOPE_GLOBAL, "enabled", use_launch_boost);
```
> `CanLaunchManagerProcessWithBoost()` function defines we want to assure that we are only providing startup boost on desktop, and that the host app's priority is at least normal. [Code Link](https://microsoft.visualstudio.com/Edge/_git/chromium.src?path=%2Fedge_embedded_browser%2Fclient%2Fwin%2Futil.cc&_a=contents&version=GBmain) Line 587

So `CreationPriority` for me is more like **"Trying to"**, we can say that in doc. 

Well, I just think about another naming `ShouldCreateWithSavinglResourceUsage` which directly shows the benefit of why setting this property.

Do you have strong preference between`ShouldCreateWithSavinglResourceUsage` and enum `CreationPriority`? I tend to use `ShouldCreateWithSavinglResourceUsage`

Reply:
Offline confirmed with Dave. The reason we choose to use enum instead of `ShouldCreateWithLowPriority` are: 
- We're not really "low priority" with applying this feature and we're like normal priority. `ShouldCreateWithLowPriority` is not accurate.
-  For  `ShouldCreateWithLowPriority`, if we want to have a real low priority feature in the future we won't know the difference. The enum give us a chance to add Low if we want to support that.

Reply:
I'm OK with ShouldCreateWithLowPriority only if we know we'll never want to make an actual low pri feature. Otherwise this is painting us into a corner